ingestion-beam/src/main/java/com/mozilla/telemetry/io/Write.java

```diff
@@ -148,6 +168,109 @@ public FileOutput(ValueProvider<String> outputPrefix, OutputFileFormat format,
     }
   }
 
+  /**
+   * Implementation of writing to local or remote files.
+   *
+   * <p>For details of the intended behavior for file paths, see:
+   * https://github.com/mozilla/gcp-ingestion/tree/master/ingestion-beam#output-path-specification
+   */
+  public static class AvroOutput extends Write {
+
+    private final ValueProvider<String> outputPrefix;
+    private final Duration windowDuration;
+    private final ValueProvider<Integer> numShards;
+    private final Compression compression;
+    private final InputType inputType;
+    private final AvroSchemaStore schemaStore;
+
+    /** Public constructor. */
+    public AvroOutput(ValueProvider<String> outputPrefix, Duration windowDuration,
+        ValueProvider<Integer> numShards, Compression compression, InputType inputType,
+        ValueProvider<String> schemasLocation) {
+      this.outputPrefix = outputPrefix;
+      this.windowDuration = windowDuration;
+      this.numShards = numShards;
+      this.compression = compression;
+      this.inputType = inputType;
+      this.schemaStore = AvroSchemaStore.of(schemasLocation);
+    }
+
+    /** A functor that constructs a GenericRecord from a PubsubMessage. */
+    public class RecordFormatter implements AvroIO.RecordFormatter<PubsubMessage> {
```

Comment:
This functor does more work than I originally thought. If there are any schema incompatibilities, the `jsonDecoder` will throw an `AvroTypeException`. I'm not sure what happens to unspecified fields, but it might be possible to stuff them into a `residue` field here. Additionally, the schemas can be modified after the fact here.

Reply:
Most of the logic here was moved into a custom JsonDecoder in #506.